kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ServerTemplate.java

```diff
@@ -176,13 +184,8 @@ public void deleteServerInstance(String serverInstanceId) {
         if (serverInstances == null) {
             return;
         }
-        Iterator<ServerInstanceKey> iterator = serverInstances.iterator();
-
-        while(iterator.hasNext()) {
-            ServerInstanceKey serverInstanceKey = iterator.next();
-            if (serverInstanceId.equals(serverInstanceKey.getServerInstanceId())) {
-                iterator.remove();
-            }
+        synchronized (lock) {
+            serverInstances.removeIf((ServerInstanceKey key) -> key.getServerInstanceId().equals(serverInstanceId));
```

Comment:
What about the other methods which modify the collections here? Namely `addServerInstance`, `addContainerSpec`, `deleteContainerSpec`? They can create this exception as well.

Reply:
what's your means?

Reply:
I mean that these 3 methods manipulate with collections as well, so in theory they can cause concurrent modification issues as well. But given that this JIRA is specifically about use case when KIE Server disconnects, it depends on you or other stakeholders if we want to go further.

Reply:
I think the issue's scope just has this. so we can keep it is simple.